op-node/rollup/derive/engine_queue.go

```diff
@@ -683,94 +680,30 @@ func (eq *EngineQueue) forceNextSafeAttributes(ctx context.Context) error {
-	if eq.isEngineSyncing() {
-		return BlockInsertTemporaryErr, fmt.Errorf("engine is in progess of p2p sync")
-	}
-	if eq.buildingID != (eth.PayloadID{}) {
-		eq.log.Warn("did not finish previous block building, starting new building now", "prev_onto", eq.buildingOnto, "prev_payload_id", eq.buildingID, "new_onto", parent)
-		// TODO: maybe worth it to force-cancel the old payload ID here.
-	}
-	fc := eth.ForkchoiceState{
-		HeadBlockHash:      parent.Hash,
-		SafeBlockHash:      eq.safeHead.Hash,
-		FinalizedBlockHash: eq.finalized.Hash,
-	}
-	id, errTyp, err := StartPayload(ctx, eq.engine, fc, attrs)
-	if err != nil {
-		return errTyp, err
-	}
-	eq.buildingID = id
-	eq.buildingSafe = updateSafe
-	eq.buildingOnto = parent
-	return BlockInsertOK, nil
+func (eq *EngineQueue) StartPayload(ctx context.Context, parent eth.L2BlockRef, attrs *AttributesWithParent, updateSafe bool) (errType BlockInsertionErrType, err error) {
+	return eq.ec.StartPayload(ctx, parent, attrs, updateSafe)
 }
 
 func (eq *EngineQueue) ConfirmPayload(ctx context.Context) (out *eth.ExecutionPayload, errTyp BlockInsertionErrType, err error) {
-	if eq.buildingID == (eth.PayloadID{}) {
-		return nil, BlockInsertPrestateErr, fmt.Errorf("cannot complete payload building: not currently building a payload")
-	}
-	if eq.buildingOnto.Hash != eq.unsafeHead.Hash { // E.g. when safe-attributes consolidation fails, it will drop the existing work.
-		eq.log.Warn("engine is building block that reorgs previous unsafe head", "onto", eq.buildingOnto, "unsafe", eq.unsafeHead)
-	}
-	fc := eth.ForkchoiceState{
-		HeadBlockHash:      common.Hash{}, // gets overridden
-		SafeBlockHash:      eq.safeHead.Hash,
-		FinalizedBlockHash: eq.finalized.Hash,
-	}
-	// Update the safe head if the payload is built with the last attributes in the batch.
-	updateSafe := eq.buildingSafe && eq.safeAttributes != nil && eq.safeAttributes.isLastInSpan
-	payload, errTyp, err := ConfirmPayload(ctx, eq.log, eq.engine, fc, eq.buildingID, updateSafe)
-	if err != nil {
-		return nil, errTyp, fmt.Errorf("failed to complete building on top of L2 chain %s, id: %s, error (%d): %w", eq.buildingOnto, eq.buildingID, errTyp, err)
-	}
-	ref, err := PayloadToBlockRef(payload, &eq.cfg.Genesis)
+	preSafe := eq.ec.SafeL2Head()
+	_, _, buildingSafe := eq.ec.BuildingPayload()
+	out, errTyp, err = eq.ec.ConfirmPayload(ctx)
 	if err != nil {
-		return nil, BlockInsertPayloadErr, NewResetError(fmt.Errorf("failed to decode L2 block ref from payload: %w", err))
+		return nil, errTyp, err
 	}
-
-	eq.unsafeHead = ref
-	eq.engineSyncTarget = ref
-	eq.metrics.RecordL2Ref("l2_unsafe", ref)
-	eq.metrics.RecordL2Ref("l2_engineSyncTarget", ref)
-
-	if eq.buildingSafe {
-		eq.pendingSafeHead = ref
-		if updateSafe {
-			eq.safeHead = ref
-			eq.postProcessSafeL2()
-			eq.metrics.RecordL2Ref("l2_safe", ref)
-		}
+	postSafe := eq.ec.SafeL2Head()
+	if buildingSafe && (preSafe != postSafe) {
```

Comment:
Would it be better to having `ConfirmPayload` return a boolean to indicate if it changed the safe head? (I could probably go either way on this but figured its worth asking).

Reply:
It'd be nice, but I wanted to limit the amount of changes I was making at once. The first attempt of changing the API at the same time as moving everything around was a lot more painful.

Reply:
Yep, makes sense - I'm ok with improving things step by step.